dm/ui/src/components/BatchImportTaskConfig/index.tsx

```diff
@@ -0,0 +1,251 @@
+      if (timer) {
+        clearInterval(timer)
+      }
+    }
+  }, [active])
+
+  if (!active) {
+    return <span>{text}</span>
+  }
+  return <span>{text + '.'.repeat(count % 4)}</span>
+}
+
+const BatchImportTaskConfig: React.FC = () => {
+  const [t] = useTranslation()
+  const [percent, setPercent] = useSafeState(0)
+  const [error, setError] = useState<any>()
+  const [importResult, setImportResult] =
+    useState<BatchImportTaskConfigResponse>()
+  const [currentStep, setCurrentStep] = useSafeState(-1)
+  const [overwrite, setOverwrite] = useState(false)
+  const [batchImportTaskConfig] = useDmapiBatchImportTaskConfigMutation()
+
+  let timerId: number
+  let stopTimer = false
+
+  const calcStepStatus = (
+    thisStep: number
+  ): { status: StepProps['status']; icon?: StepProps['icon'] } => {
+    if (currentStep === thisStep) {
+      if (currentStep === 1 && error) {
+        return { status: 'error' }
+      }
+      return { status: 'process' }
+    }
+    if (currentStep > thisStep) return { status: 'finish' }
+    return { status: 'wait' }
+  }
+
+  const handleStartImport = useCallback(() => {
+    setCurrentStep(0)
+    handleStartPlaceboProgress()
+  }, [overwrite, percent])
+
+  const increment = () => {
+    timerId = setTimeout(increment, Math.random() * 100)
+
+    setPercent(percent => {
+      if (percent >= 100) {
+        return 100
+      }
+      let n = percent
+      let amount: number
+      if (n >= 0 && n < 20) {
+        amount = 1
+      } else if (n >= 20 && n < 50) {
+        amount = 4
+      } else if (n >= 50 && n < 80) {
+        amount = 2
+      } else if (n >= 80 && n < 99) {
+        amount = 5
+      } else {
+        amount = 0
+      }
+      n = clamp(n + amount, 0, 80) // clamp here to make sure awalys 0 <= n <= 80
+      return n
+    })
+  }
+
+  const waitLittleWhileThenGoToNextStep = async () => {
+    await sleep(600) // everytime we wait for little while before continue to next step for better ux
+    setCurrentStep(c => c + 1)
+  }
+
+  const clearTimer = () => {
+    stopTimer = true
+    clearTimeout(timerId)
+  }
+
+  const handleStartPlaceboProgress = useCallback(async () => {
+    // phase 1
+    increment()
+    await sleep(1000 * Math.random() + 300)
+    setPercent(100)
+    await waitLittleWhileThenGoToNextStep()
+
+    // phase 2
+    setPercent(0)
+    try {
+      await sleep(3000)
+      const res = await batchImportTaskConfig({ overwrite }).unwrap()
+      setImportResult(res)
+    } catch (e: any) {
+      setError(e)
+      clearTimer()
+      await sleep(600)
+      setCurrentStep(4)
+      return
+    }
+
+    setPercent(100)
+    await waitLittleWhileThenGoToNextStep()
+
+    // phase 3
+    setPercent(0)
+    await sleep(1000)
+    clearTimer()
+    setPercent(100)
+
+    await waitLittleWhileThenGoToNextStep()
+  }, [percent, error])
+
+  useEffect(() => {
+    return () => {
+      if (timerId) {
+        clearTimeout(timerId)
+      }
+    }
+  }, [])
```

Comment:
interesting, how this works?

Reply:
This is a wrong way to do so and just magically works, I forgot to add `useRef` to wrap them when busy finishing the placebo progress effect.

The cleanup function in `useEffect` can access the last state of timerId (the one before the latest) just before re-render, then it will be re-declared anyway.